pypesto/startpoint/base.py

```diff
@@ -18,31 +19,189 @@ class StartpointMethod:
+
+        # assemble
+        xs = np.zeros(shape=(n_starts, dim))
+        xs[0:n_guesses, :] = x_guesses
+        xs[n_guesses:n_starts, :] = x_sampled
+
+        # check, resample and order startpoints
+        xs = self.check_and_resample(
+            xs=xs, lb=lb, ub=ub, objective=problem.objective
+        )
+
+        return xs
+
+    @abstractmethod
+    def sample(
+        self,
+        n_starts: int,
+        lb: np.ndarray,
+        ub: np.ndarray,
+    ) -> np.ndarray:
+        """Actually sample startpoints.
+
+        While in this implementation, `__call__` handles the checking of
+        guesses and resampling, this method defines the actual sampling.
+
+        Parameters
+        ----------
+        n_starts: Number of startpoints to generate.
+        lb: Lower parameter bound.
+        ub: Upper parameter bound.
+
+        Returns
+        -------
+        xs: Startpoints, shape (n_starts, n_par).
+        """
+
+    def check_and_resample(
+        self,
+        xs: np.ndarray,
         lb: np.ndarray,
         ub: np.ndarray,
         objective: ObjectiveBase,
     ) -> np.ndarray:
-        """Generate startpoints.
+        """Check sampled points for fval, grad, and potentially resample ones.
 
         Parameters
         ----------
-        n_starts:
-            Number of starts.
-        lb:
-            Lower parameter bound.
-        ub:
-            Upper parameter bound.
-        objective:
-            Objective, maybe required for evaluation.
+        xs: Startpoints candidates, shape (n_starts, n_par).
+        lb: Lower parameter bound.
+        ub: Upper parameter bound.
+        objective: Objective function, for evaluation.
 
         Returns
         -------
-        startpoints:
-            Startpoints, shape (n_starts, n_par).
+        xs:
+            Checked and potentially partially resampled startpoints,
+            shape (n_starts, n_par).
         """
+        if not self.check_fval and not self.check_grad:
+            return xs
+
+        if self.check_fval and not self.check_grad:
+            sensi_orders = 0,
+        elif not self.check_fval and self.check_grad:
+            sensi_orders = 1,
+        else:
+            sensi_orders = 0, 1
+
+        # track function values for ordering
+        fvals = np.empty(shape=(xs.shape[0],))
+
+        # iterate over all startpoint candidates
+        for ix, x in enumerate(xs):
+            # evaluate candidate
+            objective.initialize()
+            ret = objective(x, sensi_orders=sensi_orders, return_dict=True)
+            fvals[ix] = ret.get(FVAL, np.nan)
+
+            # loop until all requested sensis are finite
+            while True:
+                # discontinue if all requested sensis are finite
+                if (
+                    0 not in sensi_orders or ret[FVAL] < np.inf
+                ) and (
+                    1 not in sensi_orders or np.all(ret[GRAD] < np.inf)
```

Comment:
```suggestion
                    0 not in sensi_orders or np.isfinite(ret[FVAL])
                ) and (
                    1 not in sensi_orders or np.all(np.finite(ret[GRAD]))
```

If we accept negative infinity, we should also accept positive infinity. Or rather neither?

Reply:
On second thought, probably makes sense to leave as is, unless somebody is optimizing f(x) = inf.

Reply:
I guess requesting finite values should be fine, i.e. excluding {nan, inf, -inf}